File: tests/ten_runtime/integration/go/trigger_life_cycle_go/trigger_life_cycle_go_app/ten_packages/extension/default_extension_go/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package default_extension_go

import (
	ten "ten_framework/ten_runtime"
	"time"
)

type mainExtension struct {
	ten.DefaultExtension
}

func (p *mainExtension) OnInit(
	tenEnv ten.TenEnv,
) {
	go func() {
		// Sleep 1 second
		time.Sleep(1 * time.Second)

		// Send cmd to biz extension to check if it is started.
		cmd, _ := ten.NewCmd("check_start")
		tenEnv.SendCmd(cmd, func(te ten.TenEnv, cr ten.CmdResult, err error) {
			if err != nil {
				panic("Failed to send cmd: " + err.Error())
			}

			statusCode, _ := cr.GetStatusCode()
			if statusCode != ten.StatusCodeOk {
				panic("Failed to check start")
			}

			started, _ := cr.GetPropertyBool("started")
			if started {
				panic(
					"Biz extension should not be started as it has been set manual trigger life cycle",
				)
			}

			// Send manual trigger life cycle cmd to biz extension.
			lifeCycleCmd, _ := ten.NewTriggerLifeCycleCmd()
			lifeCycleCmd.SetStage("start")
			lifeCycleCmd.SetDests(ten.Loc{
				AppURI:        ten.Ptr(""),
				GraphID:       ten.Ptr(""),
				ExtensionName: ten.Ptr("biz"),
			})

			tenEnv.SendCmd(
				lifeCycleCmd,
				func(te ten.TenEnv, cr ten.CmdResult, err error) {
					if err != nil {
						panic("Failed to send cmd: " + err.Error())
					}

					statusCode, _ := cr.GetStatusCode()
					if statusCode != ten.StatusCodeOk {
						panic("Failed to send cmd: " + err.Error())
					}

					// Send cmd to biz extension to check if it is started.
					cmd, _ := ten.NewCmd("check_start")
					tenEnv.SendCmd(
						cmd,
						func(te ten.TenEnv, cr ten.CmdResult, err error) {
							if err != nil {
								panic("Failed to send cmd: " + err.Error())
							}

							statusCode, _ := cr.GetStatusCode()
							if statusCode != ten.StatusCodeOk {
								panic("Failed to check start")
							}

							started, _ := cr.GetPropertyBool("started")
							if !started {
								panic(
									"Biz extension should be started as it has received manual trigger life cycle cmd",
								)
							}

							tenEnv.OnInitDone()
						},
					)
				},
			)
		})
	}()
}

func (p *mainExtension) OnStop(
	tenEnv ten.TenEnv,
) {
	go func() {
		// Sleep 1 second
		time.Sleep(1 * time.Second)

		// Send cmd to biz extension to check if it is stopped.
		cmd, _ := ten.NewCmd("check_stop")
		tenEnv.SendCmd(cmd, func(te ten.TenEnv, cr ten.CmdResult, err error) {
			if err != nil {
				panic("Failed to send cmd: " + err.Error())
			}

			statusCode, _ := cr.GetStatusCode()
			if statusCode != ten.StatusCodeOk {
				panic("Failed to check stop")
			}

			stopped, _ := cr.GetPropertyBool("stopped")
			if stopped {
				panic(
					"Biz extension should not be stopped as it has been set manual trigger life cycle",
				)
			}

			// Send manual trigger life cycle cmd to biz extension.
			lifeCycleCmd, _ := ten.NewTriggerLifeCycleCmd()
			lifeCycleCmd.SetStage("stop")
			lifeCycleCmd.SetDests(ten.Loc{
				AppURI:        ten.Ptr(""),
				GraphID:       ten.Ptr(""),
				ExtensionName: ten.Ptr("biz"),
			})
			tenEnv.SendCmd(
				lifeCycleCmd,
				func(te ten.TenEnv, cr ten.CmdResult, err error) {
					if err != nil {
						panic("Failed to send cmd: " + err.Error())
					}

					statusCode, _ := cr.GetStatusCode()
					if statusCode != ten.StatusCodeOk {
						panic("Failed to send cmd: " + err.Error())
					}

					tenEnv.OnStopDone()
				},
			)
		})
	}()
}

func (p *mainExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	cmdName, _ := cmd.GetName()
	if cmdName == "test" {
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		cmdResult.SetPropertyString("detail", "ok")
		tenEnv.ReturnResult(cmdResult, nil)
	} else {
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError, cmd)
		cmdResult.SetPropertyString("detail", "unknown command")
		tenEnv.ReturnResult(cmdResult, nil)
	}
}

type bizExtension struct {
	ten.DefaultExtension

	started bool
	stopped bool
}

func (p *bizExtension) OnStart(
	tenEnv ten.TenEnv,
) {
	p.started = true

	tenEnv.OnStartDone()
}

func (p *bizExtension) OnStop(
	tenEnv ten.TenEnv,
) {
	p.stopped = true

	tenEnv.OnStopDone()
}

func (p *bizExtension) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	cmdName, _ := cmd.GetName()
	switch cmdName {
	case "check_start":
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		cmdResult.SetProperty("started", p.started)
		tenEnv.ReturnResult(cmdResult, nil)
	case "check_stop":
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		cmdResult.SetProperty("stopped", p.stopped)
		tenEnv.ReturnResult(cmdResult, nil)
	default:
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError, cmd)
		cmdResult.SetPropertyString("detail", "unknown command")
		tenEnv.ReturnResult(cmdResult, nil)
	}
}

func newDefaultExtension(name string) ten.Extension {
	if name == "main" {
		return &mainExtension{}
	}
	return &bizExtension{}
}

func init() {
	// Register addon.
	err := ten.RegisterAddonAsExtension(
		"default_extension_go",
		ten.NewDefaultExtensionAddon(newDefaultExtension),
	)
	if err != nil {
		panic("Failed to register addon.")
	}
}

File: tests/ten_runtime/integration/python/trigger_life_cycle_python/trigger_life_cycle_python_app/ten_packages/extension/default_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
import asyncio
from ten_runtime import (
    AsyncExtension,
    AsyncTenEnv,
    Cmd,
    CmdResult,
    StatusCode,
    TriggerLifeCycleCmd,
)
from ten_runtime.loc import Loc


class MainExtension(AsyncExtension):
    def __init__(self, name: str):
        super().__init__(name)

    async def on_init(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log_info("MainExtension on_init")

        await asyncio.sleep(1.0)

        # Send cmd to biz extension to check if it is started.
        cmd = Cmd.create("check_start")

        result, _ = await ten_env.send_cmd(cmd)
        assert result is not None

        started, _ = result.get_property_bool("started")
        assert (
            not started
        ), "biz extension should not be started as it has been set manual trigger life cycle"

        # Send manual trigger life cycle cmd to biz extension.
        life_cycle_cmd = TriggerLifeCycleCmd.create()
        life_cycle_cmd.set_stage("start")
        life_cycle_cmd.set_dests([Loc("", "", "biz")])

        trigger_life_cycle_result, _ = await ten_env.send_cmd(life_cycle_cmd)
        assert (
            trigger_life_cycle_result is not None
        ), "trigger_life_cycle_result is None"
        assert (
            trigger_life_cycle_result.get_status_code() == StatusCode.OK
        ), "trigger_life_cycle_result status code is not OK"

        # Send cmd to biz extension to check if it is started.
        check_start_cmd = Cmd.create("check_start")

        check_start_result, _ = await ten_env.send_cmd(check_start_cmd)
        assert check_start_result is not None, "check_start_result is None"
        assert (
            check_start_result.get_status_code() == StatusCode.OK
        ), "check_start_result status code is not OK"

    async def on_stop(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log_info("MainExtension on_stop")

        await asyncio.sleep(1.0)

        # Send cmd to biz extension to check if it is stopped.
        check_stop_cmd = Cmd.create("check_stop")

        check_stop_result, _ = await ten_env.send_cmd(check_stop_cmd)
        assert check_stop_result is not None

        stopped, _ = check_stop_result.get_property_bool("stopped")
        assert (
            not stopped
        ), "biz extension should not be stopped as it has been set manual trigger life cycle"

        # Send manual trigger life cycle cmd to biz extension.
        life_cycle_cmd = TriggerLifeCycleCmd.create()
        life_cycle_cmd.set_stage("stop")
        life_cycle_cmd.set_dests([Loc("", "", "biz")])

        trigger_life_cycle_result, _ = await ten_env.send_cmd(life_cycle_cmd)
        assert trigger_life_cycle_result is not None
        assert (
            trigger_life_cycle_result.get_status_code() == StatusCode.OK
        ), "trigger_life_cycle_result status code is not OK"

    async def on_deinit(self, ten_env: AsyncTenEnv) -> None:
        ten_env.log_info("MainExtension on_deinit")

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        cmd_name = cmd.get_name()
        ten_env.log_info(f"cmdName: {cmd_name}")

        if cmd_name == "test":
            cmd_result = CmdResult.create(StatusCode.OK, cmd)
            cmd_result.set_property_string("detail", "ok")
            await ten_env.return_result(cmd_result)
        else:
            cmd_result = CmdResult.create(StatusCode.ERROR, cmd)
            cmd_result.set_property_string("detail", "unknown command")
            await ten_env.return_result(cmd_result)


class BizExtension(AsyncExtension):
    def __init__(self, name: str):
        super().__init__(name)
        self.started: bool = False
        self.stopped: bool = False

    async def on_start(self, ten_env: AsyncTenEnv) -> None:
        self.started = True
        ten_env.log_info("BizExtension on_start")

    async def on_stop(self, ten_env: AsyncTenEnv) -> None:
        self.stopped = True
        ten_env.log_info("BizExtension on_stop")

    async def on_cmd(self, ten_env: AsyncTenEnv, cmd: Cmd) -> None:
        cmd_name = cmd.get_name()
        ten_env.log_info(f"cmdName: {cmd_name}")

        if cmd_name == "check_start":
            cmd_result = CmdResult.create(StatusCode.OK, cmd)
            cmd_result.set_property_bool("started", self.started)
            await ten_env.return_result(cmd_result)
        elif cmd_name == "check_stop":
            cmd_result = CmdResult.create(StatusCode.OK, cmd)
            cmd_result.set_property_bool("stopped", self.stopped)
            await ten_env.return_result(cmd_result)
        else:
            cmd_result = CmdResult.create(StatusCode.ERROR, cmd)
            cmd_result.set_property_string("detail", "unknown command")
            await ten_env.return_result(cmd_result)
